### PR TITLE
feat: add client keys flow

### PR DESCRIPTION
**Pull request by MatejVukosav**
- implement add-client-key endpoint
- organize code in admin folder to prevent code clutter in base service 
- rename admin mod to service mod to avoid having `admin::admin::smtng` so instead `admin::service:smtng`
- rename service function to setup to avoid `admin::service::service` but rather `admin::service::setup`
- add ApiError type to introduce error handling

POST request to http://localhost:2428/admin-api/add-client-key

ETH
```
{
    "walletSignature": "0x63f2a2c1597e7ca44c56a10ae7b040e5e7f9b8705f7d4b5c3d7582a5e76056e8193f812ae7aa2cda022a848ba750def4398d9af07c7a5cec3977c6831e20f19a1b",
    "payload": {
        "message": {
            "nodeSignature": "abcdefhgjsdajbadk",
            "clientPublicKey": "4XTTMGtEjuEpr33awoBwvXByAcTBfTHWKxvtxTHE34BbTzbk6",
            "nonce": "Y2hhbGxlbmdlMTIz",
            "applicationId": "app123",
            "timestamp": 1713320355,
            "message": "{\"nodeSignature\":\"abcdefhgjsdajbadk\",\"clientPublicKey\":\"4XTTMGtEjuEpr33awoBwvXByAcTBfTHWKxvtxTHE34BbTzbk6\"}"
        },
        "metadata": {}
    },
    "walletMetadata": {
        "type": "ETH",
        "signingKey": "0x04a4e95eeebe44a0f37b75f40957445d2d16a88c"
    }
}
```
Response
```
{
    "data": "Client key stored"
}
```

NEAR
```
{
    "walletSignature": "t5yV1R/UqEvFUfaCS3QJGQb/1WxHZsHIuBX8WixzwnQKPrYyAScQpwCd85ThrnJy3V9VIB7rXmdjrqfJA3BfCA==",
    "payload": {
        "message": {
            "clientPublicKey": "4XTTM2tQEQ7MR5HsALYB4az5NnmiVayuhL8bSSXpDZR7H2XVr",
            "nodeSignature": "abcdefhgjsdajbadk",
            "nonce": "6ZLELvCdJCIf88z0poaepcZniexsl6U2lbf5jsJwmE0=",
            "applicationId": "app123",
            "timestamp": 1713320355,
            "message": "{\"nodeSignature\":\"abcdefhgjsdajbadk\",\"clientPublicKey\":\"4XTTM2tQEQ7MR5HsALYB4az5NnmiVayuhL8bSSXpDZR7H2XVr\"}"
        },
        "metadata": {
            "recipient": "only-peers",
            "callbackUrl": "http://localhost:3000/auth/near",
            "nonce": "233,146,196,46,240,157,36,34,31,243,204,244,166,134,158,165,198,103,137,236,108,151,165,54,149,183,249,142,194,112,152,77"
        }
    },
    "walletMetadata": {
        "type": "NEAR",
        "signingKey": "ed25519:AqVajqAMLX7oaYFgHjK4Wex865bJDGeU2pCrzD6g213n"
    }
}
```
Response
```
{
    "data": "Client key stored"
}
```

In case of error response
```
{
    "error": " Challenge is too old. Please request a new challenge."
}
```

### CODE REVIEW

File: crates/server/src/lib.rs [hunk 1]
```diff
@@ -77,9 +77,8 @@ pub async fn start(
     {
         if let Some((path, handler)) = graphql::service(&config, server_sender.clone())? {
             let identity = config.identity.clone();
-            app = app
-                .route(path, handler)
-                .layer(middleware::auth::AuthSignatureLayer::new(identity));
+            app = app.route(path, handler);
+            //.layer(middleware::auth::AuthSignatureLayer::new(identity)); //TODO will be replaced with json RPC
```
Comment (MatejVukosav) [inline crates/server/src/lib.rs:81]: Authentication not working anymore as login flow has been changed.  Will update when we port to Json RPC

File: crates/identity/src/auth.rs [hunk 2]
```diff
@@ -1,19 +1,20 @@
+use eyre::eyre;
```
Comment (miraclx) [inline crates/identity/src/auth.rs]: `eyre::bail`

File: crates/identity/src/auth.rs [hunk 3]
```diff
@@ -1,19 +1,20 @@
...
 pub fn verify_peer_auth(keypair: &Keypair, msg: &[u8], signature: &[u8]) -> bool {
     keypair.public().verify(msg, signature)
 }
 
-pub fn verify_eth_signature(account: &str, message: &str, signature: &str) -> bool {
+pub fn verify_eth_signature(account: &str, message: &str, signature: &str) -> eyre::Result<bool> {
     let signature_bytes = match hex::decode(signature.trim_start_matches("0x")) {
         Ok(bytes) => bytes,
-        Err(_) => return false,
+        Err(_) => return Err(eyre!("Cannot decode signature.")),
```
Comment (miraclx) [inline crates/identity/src/auth.rs]:
```suggestion
        Err(_) => eyre::bail!("Cannot decode signature."),
```

and the others

Comment (MatejVukosav) [reply crates/identity/src/auth.rs]: I have followed the approach from `get_release` in the admin.rs but will convert everything to use bail

File: crates/server/src/admin/handlers/add_client_key.rs [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+    wallet_metadata: &WalletMetadata,
+    wallet_signature: &str,
+    payload: &Payload,
+) -> eyre::Result<bool, ApiError> {
+    match wallet_metadata.wallet_type {
+        WalletType::NEAR => {
+            let near_metadata: &NearSignatureMessageMetadata = match &payload.metadata {
+                SignatureMetadataEnum::NEAR(metadata) => metadata,
+                _ => {
+                    return Err(ApiError {
```
Comment (MatejVukosav) [inline crates/server/src/admin/handlers/add_client_key.rs:179]: Removed bail as it didn't allow to force error type in return.

File: crates/server/src/admin/handlers/add_client_key.rs [hunk 5]
```diff
@@ -0,0 +1,297 @@
+use crate::admin::service::{ApiError, ApiResponse};
+use crate::verifysignature::verify_near_signature;
+use axum::{http::StatusCode, response::IntoResponse, Json};
+use calimero_identity::auth::verify_eth_signature;
+use chrono::{Duration, TimeZone, Utc};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use tower_sessions::Session;
+use tracing::info;
```
Comment (miraclx) [inline crates/server/src/admin/handlers/add_client_key.rs]:
```suggestion
use axum::{http::StatusCode, response::IntoResponse, Json};
use calimero_identity::auth::verify_eth_signature;
use chrono::{Duration, TimeZone, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use tower_sessions::Session;
use tracing::info;

use crate::admin::service::{ApiError, ApiResponse};
use crate::verifysignature::verify_near_signature;
```

File: crates/server/src/admin/handlers/add_client_key.rs [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    metadata: Value,           // Raw JSON value for the metadata
+}
+
+fn transform_request(
+    intermediate: IntermediateAddClientKeyRequest,
+) -> eyre::Result<AddClientKeyRequest, ApiError> {
+    let metadata_enum = match intermediate.wallet_metadata.wallet_type {
+        WalletType::NEAR => {
+            let metadata = serde_json::from_value::<NearSignatureMessageMetadata>(
+                intermediate.payload.metadata.clone(),
```
Comment (miraclx) [inline crates/server/src/admin/handlers/add_client_key.rs]: I don't think clone is needed here

File: crates/server/src/admin/handlers/add_client_key.rs (hunk 4 above)
Comment (miraclx) [reply crates/server/src/admin/handlers/add_client_key.rs:179]: then change the return type to the normal `Result`

File: crates/server/src/admin/handlers/add_client_key.rs [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+        );
+
+    response
+}
+
+fn verify_node_signature(
+    wallet_metadata: &WalletMetadata,
+    wallet_signature: &str,
+    payload: &Payload,
+) -> eyre::Result<bool, ApiError> {
```
Comment (miraclx) [inline crates/server/src/admin/handlers/add_client_key.rs]:
```suggestion
) -> Result<bool, ApiError> {
```

File: crates/server/src/admin/service.rs [hunk 8]
```diff
@@ -0,0 +1,256 @@
+use super::handlers::add_client_key::add_client_key_handler;
+use crate::verifysignature;
```
Comment (miraclx) [inline crates/server/src/admin/service.rs]: v

File: crates/server/src/admin/service.rs [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+use base64::Engine;
+use near_jsonrpc_client::{methods, JsonRpcClient};
+use near_jsonrpc_primitives::types::query::QueryResponseKind;
+use near_primitives::types::{BlockReference, Finality, FunctionArgs};
+use near_primitives::views::QueryRequest;
+use rand::{thread_rng, RngCore};
+use serde::{Deserialize, Serialize};
+use serde_json::{from_slice, json};
+use std::error::Error;
+use std::fmt::{self, Display, Formatter};
```
Comment (miraclx) [inline crates/server/src/admin/service.rs]: ^

**Pull request merged**